src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/ClusterAwareConnectionProxy.java

```diff
@@ -137,6 +140,59 @@
   protected int failoverConnectTimeoutMs;
   protected int failoverSocketTimeoutMs;
 
+  /**
+   * Proxy class to intercept and deal with errors that may occur in any object bound to the current connection.
+   */
+  class JdbcInterfaceProxy implements InvocationHandler {
+    Object invokeOn;
+
+    JdbcInterfaceProxy(Object toInvokeOn) {
+      this.invokeOn = toInvokeOn;
+    }
+
+    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+      if (METHOD_EQUALS.equals(method.getName())) {
+        // Let args[0] "unwrap" to its InvocationHandler if it is a proxy.
+        return args[0].equals(this);
+      }
+
+      synchronized (ClusterAwareConnectionProxy.this) {
+        Object result = null;
+
+        try {
+          result = method.invoke(this.invokeOn, args);
+          result = proxyIfReturnTypeIsJdbcInterface(method.getReturnType(), result);
+        } catch (InvocationTargetException e) {
+          dealWithInvocationException(e);
+        } catch (IllegalStateException e) {
+          dealWithIllegalStateException(e);
+        }
+
+        return result;
+      }
+    }
+  }
+
+  /**
+   * If the given return type is or implements a JDBC interface, proxies the given object so that we can catch SQL errors and fire a connection switch.
+   *
+   * @param returnType
+   *            The type the object instance to proxy is supposed to be.
+   * @param toProxy
+   *            The object instance to proxy.
+   * @return
+   *         The proxied object or the original one if it does not implement a JDBC interface.
+   */
+  protected Object proxyIfReturnTypeIsJdbcInterface(Class<?> returnType, Object toProxy) {
```

Comment:
I think it's easier to override getNewJdbcInterfaceProxy() rather than entire proxyIfReturnTypeIsJdbcInterface()